pytorch_lightning/strategies/collaborative.py

```diff
@@ -0,0 +1,511 @@
+        for scheduler_config in self.lr_scheduler_configs:
+            scheduler_config.scheduler = HiveMindScheduler(
+                scheduler=scheduler_config.scheduler,
+                optimizer=opt,
+            )
+
+    def on_train_batch_start(self, batch: Any, batch_idx: int, dataloader_idx: int = 0) -> None:
+        if not self.hivemind_initialized:
+            self.hivemind_initialized = True
+            if self.batch_size is None:
+                try:
+                    self.batch_size = extract_batch_size(batch)
+                    log.info(f"Found per machine batch size automatically from the batch: {self.batch_size}")
+                except Exception as e:
+                    raise MisconfigurationException(
+                        "We tried to infer the batch size from the first batch of data are were unable to. "
+                        "Please provide the batch size to the Strategy (CollaborativeStrategy"
+                        f"exception raised: {e}"
+                    )
+            self._initialize_hivemind()
+
+    @property
+    def num_peers(self) -> int:
+        if self.opt:
+            return self.opt.tracker.global_progress.num_peers
+        return 1
+
+    @property
+    def dht(self) -> "hivemind.DHT":
+        return self.dht_manager.dht
+
+    @property
+    def on_tpu(self) -> bool:
+        return self.root_device.type == "xla" and _XLA_AVAILABLE
+
+    @property
+    def on_gpu(self) -> bool:
+        return self.root_device.type == "cuda" and torch.cuda.is_available()
+
+    def reduce(self, tensor: Union[Any, torch.Tensor], *args: Any, **kwargs: Any) -> Union[Any, torch.Tensor]:
+        return tensor
+
+    def all_gather(self, tensor: torch.Tensor, group: Optional[Any] = None, sync_grads: bool = False) -> torch.Tensor:
+        return tensor
+
+    @property
+    def root_device(self) -> torch.device:
+        # todo: cyclic import, someone show me the way
+        from pytorch_lightning.accelerators.cpu import CPUAccelerator
+        from pytorch_lightning.accelerators.gpu import GPUAccelerator
+
+        if isinstance(self.accelerator, GPUAccelerator):
+            return torch.device(f"cuda:{torch.cuda.current_device()}")
+        elif isinstance(self.accelerator, CPUAccelerator):
+            return torch.device("cpu")
+        raise ValueError
+
+    def model_to_device(self) -> None:
+        assert self.model is not None
+        self.model.to(self.root_device)
+
+    def setup(self, trainer: "pl.Trainer") -> None:
+        self.model_to_device()
+        super().setup(trainer)
+        if self.precision_plugin.precision in (PrecisionType.HALF, PrecisionType.MIXED):
+            self.precision_plugin.scaler = hivemind.GradScaler()
+
+    @property
+    def is_global_zero(self) -> bool:
+        return True
+
+    def barrier(self, *args: Any, **kwargs: Any) -> None:
+        pass
+
+    def broadcast(self, obj: TBroadcast, src: int = 0) -> TBroadcast:
+        return obj
```

Comment:
I think that the contract should be to sync/broadcast *if necessary*, so this should really be the strategies responsibility. However, we rely on things like this for syncrhonizing checkpointing and early stopping decisions

@SeanNaren do these work with the CollaborativeStrategy? Will it checkpoint on every process independently?

Reply:
These will not work with the collaborative strategy, and currently are orthogonal to the integration. All processes would checkpoint/early stop (since it's fully decentralized).

More importantly synchronization across processes of varying throughput isn't practical especially at the extremes!

Reply:
Understood. I think for checkpointing that's fine, however for early stopping we probably need a way to sync this as it does not make sense to only continue training for some processes (while the others quit).

Does that mean that in general there is no way to sync anything in hivemind?

Reply:
For early stopping, all processes are technically in sync so they should all stop at the same time (there is some stuff we could probably do with syncing state via the distributed hash table, but I think it's unnecessary).

the DHT is how you would sync state, however this should definitely be saved for a future PR (and is exposed for users who want to mess with it via `self.strategy.dht`!